### DISCUSSION TITLE
Bonsai python Scripting Help

### DISCUSSION THREAD

**Discussion by Michael-132-wq** [bonsai-rx/bonsai#2027] [Q&A] [answered]
Hello Bonsai Comunity, 

I am attempting to use a simple script to track a mouse in a maze. I would then like to send the x,y coordinates to a python script to calculate the distance from the center of the maze to the mouses x,y position. I have been through the documentation and am having a very difficult time understanding how to pass the x,y position from bonsai to python. 

I have attached my work flow. The center of the maze is calculated in a separate bonsai work flow and written to a CSV file. Reading that csv in python each frame would work, however if i could only read it once in bonsai and pass the maze canter x,y and mouse center x,y to the python script that would be idea. 

my over all goal is as the mouse moves in each frame the python script will recalculate the distance between the center and the mouse x,y position.

Im sorry to ask such a beginner question but have not had much luck with the documentation it feels i am missing some conceptual information 

![Screenshot 2024-09-30 111137](https://github.com/user-attachments/assets/98905aef-699f-4fdf-b950-3a06b58516ee)

**Comment by Michael-132-wq** [comment, marked as answer]
Solved: Here is a brief run down for getting python3 working with bonsai on windows 11 

1) Use CreateRuntime, no connections, leave PythonHome blank, windows will auto select the path, select your ScriptPath and put your .py file in the same folder as the bonsai program

2) Exec to execute python functions defined in your .py file. Leave Module blank and Script put function_to_run() where function_to_run() is defined in th .py file

3) ObserveOnGIL with each Set (and I would assume get) operation

*python transform is in many doc/examples but doesn't appear to be supported in the standard python scripting library (maybe in iron python? idk)

Most useful sources: Bonsai crash course -> https://www.youtube.com/watch?v=srcqJXd6Vz4
                                  Bonsai python scripting -> https://bonsai-rx.org/python-scripting/?tabs=windows
![Screenshot 2024-10-01 142504](https://github.com/user-attachments/assets/e1eda594-9814-441b-836a-85cffed0366b)

**Reply by chiyu1203** [reply]
@Michael-132-wq It is really amazing to see you got python scripting library working! I tried it before but no output returned from the Exec and I have no idea how to build up that workflow and whether certain python scripts need different bonsai modules. Could you upload an example workflow and a python file for me to try out? Many thanks in advance!

**Reply by Michael-132-wq** [reply]
*make sure your bonsai file and python script are in the same working directory 
*look for the python script output in the terminal window that pops up when you start bonsai. 

in your bonsai workflow add a create runtime node, set the script path to your script test.py

in test.py have a function run() that prints hello

in bonsai workflow add exec and point it to run(). connect exec to create runtime (doesn't NEED to be connected create runtime that but needs to be connected to something)

click start and you will see run() output hello on the bonsai terminal.

Overall bonsai is cool but their really needs to be some bare bones start up documentation. Its not intuitive and sometimes just getting hello world working (or a blinking led in embedded systems ;) is the most difficult part

**Reply by chiyu1203** [reply]
@Michael-132-wq  Thanks for the explanation! I didn't expect the output would show up on the Bonsai terminal, but finally things are clear. I got the print "hello" function working! Now I am trying to use the **Set** and **Get** nodes. **Get** is easier for me (I just need to add a variable in the script and load the name of the variable in node). However, the **Set** node  is still not clear. I thought I could change the input variable for the function (so in the case of my test workflow, returning from 1 to 2 periodically. However, when running the bonsai workflow, it first returns from 1 to 10, which is the default value I set in the python script. Then it enables the python runtime and the script in the Exec module(so returning from 1 to 5), without using the variable I put in the **Set**. This is probably something simple but I can not figure out what might be the cause. Could you give me some suggestion about that?
[python_script.zip](https://github.com/user-attachments/files/17266106/python_script.zip)
